core/src/main/scala/org/http4s/headers/Link.scala

```diff
@@ -29,49 +30,59 @@ object Link extends HeaderKey.Internal[Link] with HeaderKey.Recurring {
     import cats.parse.Parser._
 
-    // TODO: it depends on org.http4s.parser.Rfc3986Parser
-    val linkValue: Parser1[LinkValue] =
-//      rule {
-//        // https://tools.ietf.org/html/rfc3986#section-4.1
-//        (AbsoluteUri | RelativeRef) ~> { (a: Uri) =>
-//          headers.LinkValue(a)
-//        }
-//      }
-      ???
+    // https://tools.ietf.org/html/rfc3986#section-4.1
+    val linkValue: Parser[LinkValue] = {
+      import Uri._
+      uriReference(StandardCharsets.UTF_8).map { uri =>
+        headers.LinkValue(uri)
+      }
+    }
 
-    val linkValueAttr: Parser1[LinkValue] = {
-      val relParser = (linkValue ~ (string1("rel=") *> token.orElse(quotedString)))
-        .map { case (link: LinkValue, rel: String) =>
-          // https://tools.ietf.org/html/rfc8288#section-3.3
-          if (link.rel.isDefined)
-            link
-          else
-            link.copy(rel = Some(rel))
+    val linkParam: Parser[LinkParam] = {
+      val relParser = (string1("rel=") *> token.orElse(quotedString))
+        .map { rel =>
+          Rel(rel)
         }
 
-      val revParser = (linkValue ~ (string1("rev=") *> token.orElse(quotedString))).map {
-        case (link: LinkValue, rev: String) =>
-          link.copy(rev = Some(rev))
+      val revParser = (string1("rev=") *> token.orElse(quotedString)).map { rev =>
+        Rev(rev)
       }
 
-      val titleParser = (linkValue ~ (string1("title=") *> token.orElse(quotedString))).map {
-        case (link: LinkValue, title: String) => link.copy(title = Some(title))
+      val titleParser = (string1("title=") *> token.orElse(quotedString)).map { title =>
+        Title(title)
       }
 
       val typeParser = {
         val mediaRange = string1("type=") *> MediaRange.parser.orElse1(
           string1("\"") *> MediaRange.parser <* string1("\""))
-        (linkValue ~ mediaRange).map { case (link, tpe) => link.copy(`type` = Some(tpe)) }
+        mediaRange.map(tpe => Type(tpe))
       }
 
-      relParser.orElse1(revParser).orElse1(titleParser).orElse1(typeParser)
+      relParser.orElse(revParser).orElse(titleParser).orElse(typeParser)
     }
 
     val linkValueWithAttr: Parser1[LinkValue] =
-      char('<') *> linkValue <* char('>') ~ Parser.rep(char(';') *> optWs *> linkValueAttr)
+      ((char('<') *> linkValue <* char('>')) ~ Parser.rep(char(';') *> optWs *> linkParam)).map {
+        case (linkValue, linkParams) =>
+          linkParams.foldLeft(linkValue) { case (lv, lp) =>
+            lp match {
+              case Rel(rel) =>
+                if (lv.rel.isDefined) lv else lv.copy(rel = Some(rel))
+              case Rev(rev) => lv.copy(rev = Some(rev))
+              case Title(title) => lv.copy(title = Some(title))
+              case Type(tpe) => lv.copy(`type` = Some(tpe))
+            }
+          }
+      }
 
     headerRep1(linkValueWithAttr).map(links => Link(links.head, links.tail: _*))
   }
+
+  sealed trait LinkParam
```

Comment:
The old parboiled2 parser is doing some magic to manipulate the `LinkValue` that was parsed earlier.. I don't think we can express that kind of parsing with cats-parse, so I introduced this intermediary type for the link parameters, and then the `LinkValue` is augmented afterwards.

Reply:
Is this all internal to the parser?  If so, we could inline it in the parser.  (Maybe you did... I'm trying to read the indentation in the diff.)

Reply:
Nope, I need to do that :)